editor/src/components/canvas/canvas-strategies/escape-hatch-strategy.tsx

```diff
@@ -65,70 +75,134 @@ export const escapeHatchStrategy: CanvasStrategy = {
+    const parentFrame = specialSizeMeasurements?.immediateParentBounds ?? null
+    const margin = specialSizeMeasurements?.margin
     const marginPoint: LocalPoint = {
       x: -(margin?.left ?? 0),
       y: -(margin?.top ?? 0),
     } as LocalPoint
     const frameWithoutMargin = offsetRect(frame, marginPoint)
     const updatedFrame = offsetRect(frameWithoutMargin, asLocal(dragDelta ?? zeroCanvasRect))
-    return [
-      convertToAbsolute('permanent', path),
-      setCssLengthProperty(
-        'permanent',
-        path,
-        stylePropPathMappingFn('left', ['style']),
-        updatedFrame.x,
-      ),
-      setCssLengthProperty(
-        'permanent',
-        path,
-        stylePropPathMappingFn('top', ['style']),
-        updatedFrame.y,
-      ),
-      setCssLengthProperty(
-        'permanent',
-        path,
-        stylePropPathMappingFn('width', ['style']),
-        updatedFrame.width,
-      ),
-      setCssLengthProperty(
-        'permanent',
-        path,
-        stylePropPathMappingFn('height', ['style']),
-        updatedFrame.height,
-      ),
-    ]
+    const fullFrame = getFullFrame(updatedFrame)
+    const pinsToSet = filterPinsToSet(path, canvasState)
+
+    let commands: Array<CanvasCommand> = [convertToAbsolute('permanent', path)]
+    fastForEach(pinsToSet, (framePin) => {
+      const pinValue = pinValueToSet(framePin, fullFrame, parentFrame)
+      commands.push(
+        setCssLengthProperty(
+          'permanent',
+          path,
+          stylePropPathMappingFn(framePin, ['style']),
+          pinValue,
+          isHorizontalPoint(framePointForPinnedProp(framePin))
+            ? parentFrame?.width
+            : parentFrame?.height,
+        ),
+      )
+    })
+    return commands
   } else {
     return []
   }
 }
+
+function filterPinsToSet(
+  path: ElementPath,
+  canvasState: InteractionCanvasState,
+): Array<LayoutPinnedProp> {
+  const element = getElementFromProjectContents(
+    path,
+    canvasState.projectContents,
+    canvasState.openFile,
+  )
+  if (element == null) {
+    return ['top', 'left', 'width', 'height']
+  } else {
+    const horizontalProps = (['left', 'right', 'width'] as Array<LayoutPinnedProp>).filter((p) => {
+      const prop = getLayoutProperty(p, right(element.props), ['style'])
+      return isRight(prop) && prop.value != null
+    })
+    const verticalProps = (['top', 'bottom', 'height'] as Array<LayoutPinnedProp>).filter((p) => {
+      const prop = getLayoutProperty(p, right(element.props), ['style'])
+      return isRight(prop) && prop.value != null
+    })
+
+    let pinsToSet: Array<LayoutPinnedProp> = []
+    if (horizontalProps.length === 0) {
+      pinsToSet.push('left', 'width')
+    } else if (horizontalProps.length === 1) {
+      if (horizontalProps[0] !== 'width') {
+        pinsToSet.push(...horizontalProps, 'width')
+      } else {
+        pinsToSet.push('left', 'width')
+      }
+    } else {
+      pinsToSet.push(...horizontalProps)
```

Comment:
How does this behave if more than 2 props are set in each direction? Does it update them all correctly, and then leave it to the browser to determine which prop to ignore?

Reply:
For more than 2 props set per direction all of them should be updated, I need to cover this part with more tests.